tests/engine/cloud/test_cloud_flow_runner.py

```diff
@@ -381,6 +385,49 @@ def is_p_three(p):
     assert last_state.cached_inputs["p"] == exp_res
 
 
+def test_task_failure_with_upstream_secrets_doesnt_store_secret_value_and_recompute_if_necessary(
+    client
+):
+    @prefect.task(max_retries=2, retry_delay=timedelta(seconds=100))
+    def is_p_three(p):
+        if p == 3:
+            raise ValueError("No thank you.")
+        return p
+
+    with prefect.Flow("test") as f:
+        p = prefect.tasks.secrets.Secret("p")
+        res = is_p_three(p)
+
+    with prefect.context(secrets=dict(p=3)):
+        state = CloudFlowRunner(flow=f).run(return_tasks=[res])
+
+    assert state.is_running()
+    assert isinstance(state.result[res], Retrying)
+
+    exp_res = Result(3, result_handler=SecretResultHandler(p))
+    assert not state.result[res].cached_inputs["p"] == exp_res
+    exp_res.store_safe_value()
+    assert state.result[res].cached_inputs["p"] == exp_res
+
+    ## here we set the result of the secret to a saferesult, ensuring
+    ## it will get converted to a "true" result
+    ## moreover, we make the safe result value 3 and we expect
+    ## that the upstream value will actually get recomputed from context
+    ## through the SecretResultHandler
+    safe = SafeResult(3, result_handler=SecretResultHandler(p))
```

Comment:
In reality, wouldn't the SafeResult be `p` since that's the secret name, rather than the secret value?

Reply:
Yea, that is true - I was just trying to stack the cards in favor of the value "3" just in case -- I can change if you'd prefer

Reply:
I hate to nitpick on this but since I view this test as "enshrining" the expected high level API, i think it's important to be clear that the value isn't part of the expected payload.

Reply:
No worries at all - so would you like me to mention this in the comment or just convert it to the safe result we'd expect in practice?

Reply:
My preference would be change it to `SafeResult('p'...)`

Reply:
👍 updated